File: lista_4/dpp.py
import threading
import time
import random

N = 5  # number of philosophers (and forks)
k = 3  # number of meals each philosopher needs to eat

class State:
    THINKING = 0
    HUNGRY = 1
    EATING = 2

def left(i):
    return (i - 1 + N) % N

def right(i):
    return (i + 1) % N

state = [State.THINKING] * N
meal_count = [0] * N  # array to track the number of meals each philosopher has eaten
critical_region_mtx = threading.Lock()
output_mtx = threading.Lock()

both_forks_available = [threading.Semaphore(0) for _ in range(N)]

def my_rand(min_val, max_val):
    return random.randint(min_val, max_val)

def test(i):
    if state[i] == State.HUNGRY and state[left(i)] != State.EATING and state[right(i)] != State.EATING:
        state[i] = State.EATING
        both_forks_available[i].release()

def think(i):
    duration = my_rand(400, 800) / 1000.0
    with output_mtx:
        print(f"{i} is thinking {duration * 1000:.0f}ms")
    time.sleep(duration)

def take_forks(i):
    with critical_region_mtx:
        state[i] = State.HUNGRY
        with output_mtx:
            print(f"\t\t\t{i} is HUNGRY")
        test(i)
    both_forks_available[i].acquire()

def eat(i):
    duration = my_rand(400, 800) / 1000.0
    with output_mtx:
        print(f"\t\t\t\t\t\t{i} is eating {duration * 1000:.0f}ms his {meal_count[i]+1}. meal")
    time.sleep(duration)
    meal_count[i] += 1

def put_forks(i):
    with critical_region_mtx:
        state[i] = State.THINKING
        test(left(i))
        test(right(i))

def philosopher(i):
    while meal_count[i] < k:
        think(i)
        take_forks(i)
        eat(i)
        put_forks(i)
    with output_mtx:
        print(f"\t\t\t\t\t\t\t\t\t\t\tPhilosopher {i} has finished eating {k} meals.")

if __name__ == "__main__":
    print("dp_14")

    threads = [threading.Thread(target=philosopher, args=(i,)) for i in range(N)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    
    print("All philosophers have eaten their meals.")

File: lista_4/dpp.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

type Philosopher struct {
	number    int
	chopstick chan bool
	neighbor  *Philosopher
}

func makePhilosopher(number int, neighbor *Philosopher) *Philosopher {
	phil := &Philosopher{number, make(chan bool, 1), neighbor}
	phil.chopstick <- true
	return phil
}

func (phil *Philosopher) think() {
	fmt.Printf("%d is thinking\n", phil.number)
	time.Sleep(time.Duration(rand.Int63n(1e9)))
}

func (phil *Philosopher) eat(mealNum int, wg *sync.WaitGroup) {
	fmt.Printf("\t\t\t\t\t\t%d is eating his %d. meal\n", phil.number, mealNum)
	time.Sleep(time.Duration(rand.Int63n(1e9)))
	wg.Done()
}

func (phil *Philosopher) getChopsticks() {
	timeout := make(chan bool, 1)
	go func() { time.Sleep(1e9); timeout <- true }()
	<-phil.chopstick
	fmt.Printf("\t\t\t%d is HUNGRY\n", phil.number)
	select {
	case <-phil.neighbor.chopstick:
		//fmt.Printf("\t\t\t\t\t\t%d is eating\n", phil.number)
		//fmt.Printf("\t\t\t\t\t\t%d is eating his 1. meal\n", phil.number)
		return
	case <-timeout:
		phil.chopstick <- true
		phil.think()
		phil.getChopsticks()
	}
}

func (phil *Philosopher) returnChopsticks() {
	phil.chopstick <- true
	phil.neighbor.chopstick <- true
}

func (phil *Philosopher) dine(announce chan *Philosopher, k int) {
	var wg sync.WaitGroup
	for j := 0; j < k; j++ {
		phil.think()
		phil.getChopsticks()
		wg.Add(1)
		phil.eat(j+1, &wg)
		phil.returnChopsticks()
	}
	wg.Wait()
	announce <- phil
}

func main() {
	n := 5 // liczba filozofów
	k := 3 // liczba posiłków dla każdego filozofa
	philosophers := make([]*Philosopher, n)
	var phil *Philosopher
	for i := 0; i < n; i++ {
		phil = makePhilosopher(i, phil)
		philosophers[i] = phil
	}
	philosophers[0].neighbor = phil
	fmt.Printf("There are %d philosophers sitting at a table.\n", n)
	fmt.Printf("Each philosopher will eat %d meals.\n", k)
	announce := make(chan *Philosopher)
	for _, phil := range philosophers {
		go phil.dine(announce, k)
	}
	for i := 0; i < n; i++ {
		phil := <-announce
		fmt.Printf("\t\t\t\t\t\t\t\t\t\t\tPhilosopher %d has finished eating %d meals.\n", phil.number, k)
	}
	fmt.Printf("All philosophers have eaten their meals.\n")
}
